File: Python3/Hard/WrongAnswer/decode-ways-ii_3.py
class Solution:
    def numDecodings(self, s: str) -> int:
        
        # dp[i] = num of ways to decode s[:i]
        # dp[i] = dp[i-1] * (1 if s[i-1] in '123456789' else 9) + dp[i-2] * (9 if s[i-2] == '1' else 6 if s[i-2] == '2' else 15 if s[i-2] == '*' else 0)
        # dp[0] = 1
        # dp[1] = 9 if s[0] == '*' else 1 if s[0] in '123456789' else 0
        MOD = 10**9 + 7
        dp = [0] * (len(s) + 1)
        dp[0] = 1
        dp[1] = 9 if s[0] == '*' else 1 if s[0] in '123456789' else 0
        for i in range(2, len(s) + 1):
            if s[i-1] in '123456789':
                dp[i] += dp[i-1]
            else:
                dp[i] += dp[i-1] * 9
            if s[i-2] == '1':
                dp[i] += dp[i-2]
            elif s[i-2] == '2':
                dp[i] += dp[i-2] * 6
            elif s[i-2] == '*':
                dp[i] += dp[i-2] * 15
            dp[i] %= MOD
        return dp[-1]

File: C++/Hard/WrongAnswer/decode-ways-ii_3.cpp
class Solution {
public:
    int numDecodings(string s) {
        
        int n = s.size();
        const int mod = 1e9 + 7;
        vector<long long> f(n + 1);
        f[0] = 1;
        for (int i = 1; i <= n; i++) {
            if (s[i - 1] == '0') {
                if (i == 1 || s[i - 2] == '0' || s[i - 2] >= '3') {
                    return 0;
                }
                f[i] = f[i - 2];
            } else if (s[i - 1] == '*') {
                f[i] = 9 * f[i - 1];
                if (i > 1) {
                    if (s[i - 2] == '1') {
                        f[i] = (f[i] + 9 * f[i - 2]) % mod;
                    } else if (s[i - 2] == '2') {
                        f[i] = (f[i] + 6 * f[i - 2]) % mod;
                    } else if (s[i - 2] == '*') {
                        f[i] = (f[i] + 15 * f[i - 2]) % mod;
                    }
                }
            } else {
                f[i] = f[i - 1];
                if (i > 1 && s[i - 2] != '0') {
                    int ten = (s[i - 2] - '0') * 10;
                    int one = s[i - 1] - '0';
                    if (ten + one <= 26) {
                        f[i] = (f[i] + f[i - 2]) % mod;
                    }
                }
            }
        }
        return f[n];
    }
};